File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    from collections import deque

    def check(arr):
        stack = []
        for i in arr:
            if i in '([{':
                stack.append(i)
            else:
                if not stack:
                    return 0
                top = stack.pop()
                if top == '(' and i != ')':
                    return 0
                if top == '[' and i != ']':
                    return 0
                if top == '{' and i != '}':
                    return 0
                    
        if stack:
            return 0

        return 1
    
    
    answer = 0
    q = deque(s)
    L = len(s)
    for i in range(L):
        if check(q):
            answer += 1
        q.rotate(-1)
        
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>
#include <algorithm>
using namespace std;


bool check(string str) {
    stack<char> s;
    for(char now : str) {
        if(now == '[' || now == '{' || now == '(') {
            s.push(now);
        } else {
            if(s.empty()) return false;
            if(s.top() == '[' && now != ']') return false;
            if(s.top() == '{' && now != '}') return false;
            if(s.top() == '(' && now != ')') return false;
            s.pop();
        }
    }
    return s.empty();
}

int solution(string s) {
    int answer = 0;
    string rotated = s;
    for(int i = 0; i < s.size(); i++) {
        if(check(rotated)) answer++;
        rotate(rotated.begin(), rotated.begin() + 1, rotated.end());
    }
    return answer;
}
